components/logins/src/db.rs

```diff
@@ -373,33 +348,93 @@ impl LoginDb {
-        }
+        assert_eq!(
+            rows_changed, 1,
+            "adding a new record must have changed 1 row"
+        );
+        Ok(())
+    }
+
+    fn update_existing_login(&self, login: &Login) -> Result<()> {
+        // assumes the "local overlay" exists, so the guid must too.
+        let sql = format!(
+            "UPDATE loginsL
+             SET local_modified      = :now_millis,
+                 timeLastUsed        = :time_last_used,
+                 timePasswordChanged = :time_password_changed,
+                 httpRealm           = :http_realm,
+                 formActionOrigin    = :form_action_origin,
+                 usernameField       = :username_field,
+                 passwordField       = :password_field,
+                 timesUsed           = :times_used,
+                 encFields           = :enc_fields,
+                 origin              = :origin,
+                 -- leave New records as they are, otherwise update them to `changed`
+                 sync_status         = max(sync_status, {changed})
+             WHERE guid = :guid",
+            changed = SyncStatus::Changed as u8
+        );
+
+        self.db.execute_named(
+            &sql,
+            named_params! {
+                ":origin": login.fields.origin,
+                ":http_realm": login.fields.http_realm,
+                ":form_action_origin": login.fields.form_action_origin,
+                ":username_field": login.fields.username_field,
+                ":password_field": login.fields.password_field,
+                ":time_last_used": login.time_last_used,
+                ":times_used": login.times_used,
+                ":time_password_changed": login.time_password_changed,
+                ":enc_fields": login.enc_fields,
+                ":guid": &login.id,
+                // time_last_used has been set to now.
+                ":now_millis": login.time_last_used,
+            },
+        )?;
+        Ok(())
+    }
+
+    pub fn add(&self, login: UpdatableLogin, encdec: &EncryptorDecryptor) -> Result<Login> {
+        let guid = Guid::random();
+        let now_ms = util::system_time_ms_i64(SystemTime::now());
+
+        let (new_fields, new_enc) =
+            self.fixup_and_check_for_dupes(&guid, login.fields, login.enc_fields, &encdec)?;
+        let result = Login {
+            id: guid.to_string(),
+            fields: new_fields,
+            enc_fields: new_enc.encrypt(&encdec)?,
+            time_created: now_ms,
+            time_password_changed: now_ms,
+            time_last_used: now_ms,
+            times_used: 1,
+        };
+        let tx = self.unchecked_transaction()?;
+        self.insert_new_login(&result)?;
         tx.commit()?;
-        Ok(login)
+        Ok(result)
     }
 
-    pub fn import_multiple(&self, logins: &[Login]) -> Result<MigrationMetrics> {
+    pub fn import_multiple(
+        &self,
+        logins: Vec<Login>,
```

Comment:
stuff got a but messy because this public function wants to add new records via a `Login` struct rather than an `UpdatableLogin` (eg, `validate_and_fixup` for a `Login` struct is only needed because this exists). This was the thing that pushed me over the edge of creating the `LoginFields` struct rather than flattening them.

Reply:
I'm not sure what you are asking here, but to restate what you said, validating a `Login` does just validate the `fields` and `enc_fields` members. This means there's no validation on the metadata, which possibly should happen, but the existing code didn't check that either IIUC.